### ISSUE TITLE
Incorrect CS8604 warning when using [MaybeNull] with unconstrained generics

### ISSUE THREAD

**Issue by mbashov** [dotnet/roslyn#43594]
**Version Used**: 
.NET Core SDK 3.1.201
Microsoft (R) Visual C# Compiler version 3.5.0-beta4-20153-05 (20b9af91)

Apparently, the 3.6.0-3 compiler also behaves like this.

**Steps to Reproduce**: pass an object received from a generic `[return: MaybeNull]` method to another generic method:
```csharp
#nullable enable
using System.Diagnostics.CodeAnalysis;
public static class C {
    [return: MaybeNull]
    static T M1<T>() => default!;
    
    static void M2<T>(T arg) {}
    
    static void M<T>()
    {
        M2(M1<T>());
    }
}
```

**Expected Behavior**: no warning, as in the old 3.4.1 compiler from .NET Core SDK 3.1.103

**Actual Behavior**:
```
warning CS8604: Possible null reference argument for parameter 'arg' in 'void C.M2<T>(T arg)'.
```

[sharplab](https://sharplab.io/#gist:1d899118847ce3b0e30c586ebba59358) - in "Default" mode the warning is not generated, but it is generated for the master branch

**Comment by jcouv** [comment]
This will likely be covered by `T??` work. Tracked by https://github.com/dotnet/roslyn/issues/29146
